+str #16885 add splitAfter

ktoso opened the pull request:
Resolves #16885 

It's very similar to splitWhen, but different enough to warant it's own class I think. Sharing code would end up being more confusing than worth it I think.

@jrudolph please sanity check it has the semantics you need.

ktoso committed:
+str #16885 add splitAfter
Implementation is shared with splitWhen - see Split / SplitWhere
Allows for easy addition of splitWhen(x => Decision) if we'd like to
Resolves #16885

ktoso commented on akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:
```diff
@@ -487,11 +487,46 @@ class Source[+Out, +Mat](delegate: scaladsl.Source[Out, Mat]) extends Graph[Sour
+   * is [[akka.stream.Supervision.Resume]] or [[akka.stream.Supervision.Restart]]
+   * the element is dropped and the stream and substreams continue.
+   *
+   * See also [[Source.splitAfter]].
```
javadsl.Source had a bit of missing scaladoc, fixed. The rest is all consistent.

drewhk commented on akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:
```diff
@@ -487,11 +487,46 @@ class Source[+Out, +Mat](delegate: scaladsl.Source[Out, Mat]) extends Graph[Sour
+   * is [[akka.stream.Supervision.Resume]] or [[akka.stream.Supervision.Restart]]
+   * the element is dropped and the stream and substreams continue.
+   *
+   * See also [[Source.splitWhen]].
```
Please now also add the summary section that is used in all scaladoc entries, see: https://github.com/akka/akka/pull/17164

ktoso replied:
oh, sure!

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/SplitAfterProcessorImpl.scala:
```diff
@@ -0,0 +1,94 @@
+  def props(settings: ActorFlowMaterializerSettings, splitPredicate: Any ⇒ Boolean): Props =
+    Props(new SplitAfterProcessorImpl(settings, splitPredicate))
+
+  private trait SplitDecision
```
Hmm, what if we add a SplitBefore and SplitAfter, change the function from Any => Boolean to Any => SplitWhere, then we have one unified implementation for splitWhen and splitAfter (which just emit Continue or SplitBefore/SplitWhen depending on the predicate), and also different other splits.

ktoso replied:
Sounds neat, I'll give it a shot.

drewhk replied:
time-box it, it just seemed straightforward.

ktoso replied:
So it's here on [wip-one-impl-to-rule-them-all-ktoso](https://github.com/ktoso/akka/commit/9fabf3cc35471a91642473a1aac50f9e47953850).
Though I think it's gotten more complicated (2 separate modes, mixted together) and more allocations. 

Could be cleaned up a bit, but first wanted feedback if we prefer one impl or two simple ones. Please chime in guys @drewhk @rkuhn @patriknw (2 simple impls is this PR, 1 smart impl is the above commit).

patriknw replied:
I'm not sure I see the added complexity. It doesn't look much compared to the two different classes.

drewhk replied:
I agree, one less class to maintain.

ktoso replied:
Cool, last question then - with the new class we're able to expose: `def split(Out => SplitDecision)` should we then expose `splitWhen`, `splitAfter` and `split` (split can be used fro dropping elements at will and SplitDrop as well).

2 new methods API wise, but seems pretty powerful, sounds good?

patriknw replied:
if we go with two methods `splitWhen` and `splitAfter` I don't think we should also expose `split` with the enums. Otherwise only `split`, but the enums are not so nice (esp for java).

ktoso replied:
Ok, so I'll do this: 

* Use the 1 class impl; It's theoretically more powerful than we're exposing - just splitWhen and splitAfter. This solves the immediate ticket. 
* I'll open a ticket to decide if we want to expose `split(A => Decision)` or something else like it (or nothing).

ktoso replied:
Pushed `SplitWhere` based impls to this PR.

patriknw commented on akka-stream/src/main/scala/akka/stream/impl/SplitAfterProcessorImpl.scala:
```diff
@@ -0,0 +1,94 @@
+  private def decideSplit(elem: Any): SplitDecision =
+    try {
+      val predicate = splitPredicate(elem)
+      if (predicate) Split else Continue
```
the `val predicate` is only used once, so `if (splitPredicate(elem)) Split else Continue`?

ktoso replied:
thanks :)

jrudolph commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -390,11 +392,38 @@ class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Graph
+   * three substreams will be produced with lengths 1, 2, and 3:
+   *
+   * {{{
+   * true,               // element goes into first substream
```
Maybe, also add one or more leading `false` values here, to make it sure that they are included in the first stream.

jrudolph commented:
You should probably also search for and look at the previous suggestion(s) that @sirthias made regarding generalizing the functionality and how the APIs could look like. I couldn't find them quickly but they are probably still there somewhere in the issues.

patriknw commented on akka-stream/src/main/scala/akka/stream/impl/SplitWhereProcessorImpl.scala:
```diff
@@ -0,0 +1,118 @@
+
+import scala.util.control.NonFatal
+
+object SplitDecision {
```
internal

ktoso replied:
Ah yes, thanks.

ktoso commented on akka-stream/src/main/scala/akka/stream/impl/SplitWhereProcessorImpl.scala:
```diff
@@ -0,0 +1,118 @@
+  case object SplitAfter extends SplitDecision
+
+  /** Drops the current element, completes the current substream. The next element will be emitted into a new substream. */
+  case object SplitDrop extends SplitDecision
```
Not implemented, I'll remove

patriknw commented:
LGTM

ktoso commented:
The failures are real, in `ResponseParserSpec` which uses a splitWhen... I thought it's "when first element matched" semantics changes, but I think those are OK (needed to add an "first element flag" otherwise I'd emit an empty substream if first element matched on splitWhen). Current semantics:

```
splitAfter: List(false, true, true, false, false, true, false)
1-false
1-true
———————————
2-true
———————————
3-false
3-false
3-true
———————————
4-false


splitWhen: List(false, true, true, false, false, true, false)
5-false
———————————
6-true
———————————
7-true
7-false
7-false
———————————
8-true
8-false


splitAfter: List(true, false, true, true, false, false, true, false)
9-true
10-false
———————————
10-true
———————————
11-true
12-false
12-false
———————————
12-true
13-false


splitWhen: List(true, false, true, true, false, false, true, false)
14-true    // technically not obvious splitWhen == SplitBefore => Source() + Source(14-true)
———————————
15-false
———————————
16-true
———————————
17-true
17-false
17-false
———————————
18-true
18-false
```

Need to look more into it (failures introduced after collapsing the two impls into one). I still feel having them separate was much simpler to deal with.

jrudolph commented:
One thing these kind of diagrams are not showing is when the substreams are closed. In the case of `splitWhen` it's in the moment when the next substream is opened (or the input stream is completed). In the case of `splitAfter` it's after the last element, and until the next element arrives there's no substream open.

Not sure if it's related to the failure, I haven't yet looked at the failure or the implementation in detail.

jrudolph commented:
The error in `FlowWithHeadAndTail` is a sign that it could be related to completion: our `headAndTail` cannot deal with empty substreams so it seems empty substreams are now introduced somewhere where they didn't occur before.

jrudolph commented:
Btw. as I didn't mentioned it in the original ticket: the motivation for `splitAfter` from our side is exactly the changed completion behavior: with `splitWhen` we needed to introduce an otherwise superfluous message `MessageEnd` to close the previous substream eagerly. With `splitAfter` we could clean up the event protocol to let the last actual data event close the substream automatically.

ktoso commented:
Thanks for the feedback @jrudolph (and sorry I did not reply here sooner). Yes I think it's empty streams, I'll look at completion behaviour so that it matches your need (add tests for it).

patriknw commented:
what is the status here?

ktoso commented:
Status here is I need to look over the failures in http (akka-stream-tests-experimental is green), possibly small change in splitWhen semantics, and when fixed should be good to go.

ktoso commented:
This will slip from RC1, there seems to be some semantical change other than the one I just changed between the old and new impl of splitWhen. Don't want to break HTTP with pulling this in for RC1.

ktoso commented:
rebased, fixed and pushed :-)
The bug was related to the first element in a `splitWhen` causing an empty substream, which was *not* the behaviour of splitWhen previously (caused by forgetting to set a boolean flag in one spot).

ktoso commented:
Failure was SSL spec :cry:

ktoso commented:
PLS BUILD

ktoso commented:
@drewhk @2m @jrudolph this needs a last review and should be merged I think.

drewhk commented on akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:
```diff
@@ -509,11 +509,72 @@ class Source[+Out, +Mat](delegate: scaladsl.Source[Out, Mat]) extends Graph[Sour
+   * is [[akka.stream.Supervision.Resume]] or [[akka.stream.Supervision.Restart]]
+   * the element is dropped and the stream and substreams continue.
+   *
+   * '''Emits when''' an element for which the provided predicate is true, opening and emitting a new substream for subsequent element
```
add to the table in the docs

ktoso replied:
done, thanks for reminding me!

ktoso commented:
Final LGTM?

drewhk commented:
merge

ktoso commented:
Merged `splitAfter` // cc @jrudolph might be interesting for you for websockets :)

jrudolph commented:
Thanks, @ktoso. Will try when when implementing the client-side of websocket support.

ktoso commented:
Sounds great :)

hepin1989 commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -803,8 +812,47 @@ trait FlowOps[+Out, +Mat] {
+  def splitWhen[U >: Out](p: Out ⇒ Boolean): Repr[Out, Mat]#Repr[Source[U, Unit], Mat] = {
+    val f = p.asInstanceOf[Any ⇒ Boolean]
+    withAttributes(name("splitWhen")).andThen(Split(el ⇒ if (f(el)) SplitBefore else Continue))
+  }
```
Why not
```
  def splitWhen[U >: Out](p: Out ⇒ Boolean): Repr[Source[U, Unit], Mat] = {
    val f = p.asInstanceOf[Any ⇒ Boolean]
    andThen(Split(el ⇒ if (f(el)) SplitBefore else Continue, name("splitWhen")))
  }
```
?

ktoso replied:
nice one, would you like to submit that change as PR? :)

ktoso replied:
And on https://github.com/akka/akka/pull/17196/files#diff-d25227230bef7fd44d678521a8f07f1cR854 as well

hepin1989 replied:
then how about adding this:
```
  object Split{
    def when(f:Any ⇒ Boolean)= {
      Split(el ⇒ if (f(el)) SplitAfter else Continue, name("splitAfter"))
    }
    
    def after(f:Any ⇒ Boolean)= {
      Split(el ⇒ if (f(el)) SplitAfter else Continue, name("splitAfter"))
    }
  }
```
and then could have:
```
  def groupBy[K, U >: Out](f: Out ⇒ K): Repr[(K, Source[U, Unit]), Mat] =
    andThen(GroupBy(f.asInstanceOf[Any ⇒ Any]))

  def splitWhen[U >: Out](p: Out ⇒ Boolean): Repr[Source[U, Unit], Mat] = {
    andThen(Split.when(p.asInstanceOf[Any ⇒ Boolean]))
  }
```
?

ktoso replied:
That's a nice refactoring, let's do it :) please submit a PR
